In [12]:
import pandas as pd
from googleapiclient.discovery import build
import numpy as np

Використовуємо бібліотеку googleapiclient.discovery для збору інформації про 100 (може бути інша кількість) найпопулярніших відео з певної країни (не України) в Youtube 

In [3]:
def get_the_most_popular_videos(api_key, max_results=100):
    youtube = build("youtube", "v3", developerKey=api_key)

    #список для зберігання деталей відео
    videos = []

    request = youtube.videos().list(
        part='snippet,contentDetails,statistics',
        chart='mostPopular',
        regionCode='MA',
        maxResults=100
    )

    while request and len(videos) < max_results:
        response = request.execute()
        for item in response['items']:
            video_details = {
                'video_id': item['id'],
                'title': item['snippet']['title'],
                'description': item['snippet']['description'],
                'published_at': item['snippet']['publishedAt'],
                'channel_id': item['snippet']['channelId'],
                'channel_title': item['snippet']['channelTitle'],
                'category_id': item['snippet']['categoryId'],
                'tags': item['snippet'].get('tags', []),
                'duration': item['contentDetails']['duration'],
                'definition': item['contentDetails']['definition'],
                'caption': item['contentDetails'].get('caption', 'false'),
                'view_count': item['statistics'].get('viewCount', 0),
                'like_count': item['statistics'].get('likeCount', 0),
                'dislike_count': item['statistics'].get('dislikeCount', 0),
                'favorite_count': item['statistics'].get('favoriteCount', 0),
                'comment_count': item['statistics'].get('commentCount', 0)
            }
            videos.append(video_details)



    return videos[:max_results]

In [4]:
def create_csv(data, file):
    df = pd.DataFrame(data)
    df.to_csv(file)

In [8]:
API_KEY = 'AIzaSyB8zKWIAnRBtW3IdDKnYJFcgWQNnWGQN1E'
videos_list = get_the_most_popular_videos(API_KEY)
create_csv(videos_list, "the_most_popular_videos.csv")

In [9]:
the_most_popular_videos = pd.read_csv("the_most_popular_videos.csv")
the_most_popular_videos

Unnamed: 0     video_id  \
0            0  7aLqIpxvny0   
1            1  el-20sun0UA   
2            2  1dBGZUoqbjo   
3            3  Ux2GHC9QXnM   
4            4  NMKj0BVJono   
..         ...          ...   
95          95  Abi0w98MaEw   
96          96  Jj25RDuBc2g   
97          97  6nsxsHVLbco   
98          98  Gf9RlABpwRU   
99          99  WOvtPpQSpYE   

                                                title  \
0                                         حفلة عرسي 💍   
1   Vlog- هادي ماكنتش دايراها فالحسبان 😪 ماما تصدم...   
2   أستاذة تدافع عن  سمعة زميلتها الراحلة هاجر وتك...   
3             كان غادي يبكي بالفرحة ملي شاف الموطور🥺🔥   
4               DUKE - WALIDIA (Official Music Video)   
..                                                ...   
95        تحدي الحقيبة مع ابوعمر (الحلقة الختامية)!🧳🔥   
96  ماين كرافت لاست لايف #14 اندرمان فارم ( اضخم م...   
97  مقلب المقابلة التلفزيونية في أعضاء TU 🤣🔥 | مقا...   
98  GRAVITY UPDATE TRAILER + FRUIT SHOWCASE | Blox...   
99  طبخت وجبة كاملة بس باستخدام أدوات تصليح | كارث...   

                                          description          published_at  \
0   حسابي على الانستغرام : http://instagram.com/na...  2025-04-15T19:52:34Z   
1   حسابي الجديد بانستغرام : \nhttps://www.instagr...  2025-04-16T19:55:21Z   
2   مدار21 جريدة إلكترونية مغربية مستقلة   \n\n\nا...  2025-04-15T19:40:36Z   
3                                                 NaN  2025-04-16T20:18:41Z   
4   DUKE - WALIDIA (Official Music Video)\n\nSubsc...  2025-04-16T19:41:31Z   
..                                                ...                   ...   
95  متجر تويلف ولا تنسى كود الخصم (ANY)😍\nhttps://...  2025-04-13T10:00:24Z   
96  ما عليك سوى الانضمام إلى عضوية هذه القناة للاس...  2025-04-15T10:00:26Z   
97  ► قناتي الثانية حياك عزيزي لا تستحي: \nhttps:/...  2025-04-12T15:05:00Z   
98  UPDATE RELEASE DATE: https://www.roblox.com/ev...  2025-04-12T03:59:06Z   
99  في فيديو اليوم طبخت وجبة باستخدام ادوات غريب !...  2025-04-13T16:05:25Z   

                  channel_id                   channel_title  category_id  \
0   UCWXYdNyGX01rnTB-tQci8JQ                   Narins Beauty           24   
1   UCMAia3HmonZbo1nqx-4ofNA  TAYTMATERS- صوفيا نهيلة عيشوشة           22   
2   UCAtDSTJxS8ECR-Fo3cujV6Q                     Madar21مدار           22   
3   UCAc1wRfZrcI5tus55j7BTog              Oussama Copenhagen           24   
4   UClAbH5V05PuIJMwOkW25xJw                   Duke Officiel           10   
..                       ...                             ...          ...   
95  UCw0rwoFG21ig-51NbpzJSkw                     abu_abeer16           20   
96  UC3CFqU4KlktVPPJ1Km87pRg                 مصطفى GAME OVER           20   
97  UC0fkyMAE7T3OFs7lnGCr33A                       توبز TOPZ           24   
98  UC_xlDCBWQPlihORIZLWtpbA                     Gamer Robot           20   
99  UCmWV4bvAO4vUQtRZayT1usw                          Ayatee           26   

                                                 tags  duration definition  \
0   ['Narıns beauty', 'نارين', 'تحديات شيرين و نار...  PT32M23S         hd   
1                                                  []  PT14M42S         hd   
2                                                  []   PT6M13S         hd   
3                                                  []  PT10M59S         hd   
4   ['Duke', 'ديوك', 'Walidia', 'والديا', 'Ana li ...    PT3M7S         hd   
..                                                ...       ...        ...   
95  ['فيفا', 'ابو عبير', 'بكجات', 'طحن بدون شحن', ...   PT52M2S         hd   
96                        ['ماين كرافت', 'minecraft']  PT32M43S         hd   
97  ['توبز', 'topz', 'TOPZ', 'اشياء', 'تفجع', 'اشي...  PT31M22S         hd   
98  ['blox fruits', 'admin', 'zioles', 'blox fruit...  PT19M56S         hd   
99  ['طبخت وجبة كاملة باستخدام أدوات تصليح', 'Aya'...  PT16M31S         hd   

    caption  view_count  like_count  dislike_count  favorite_count  \
0     False     6196118      421765              0               

In [11]:
pd.set_option("display.max.rows", 100)
the_most_popular_videos

Unnamed: 0     video_id  \
0            0  7aLqIpxvny0   
1            1  el-20sun0UA   
2            2  1dBGZUoqbjo   
3            3  Ux2GHC9QXnM   
4            4  NMKj0BVJono   
5            5  jhgNSBzzvzQ   
6            6  h7IeZ4qTdFo   
7            7  lrxRJboEGrw   
8            8  WKB2IQz-rcs   
9            9  kCgV4Bl72mY   
10          10  oVgEgVuK2Y8   
11          11  oXzjaCjxcIs   
12          12  y4ZqKsTQN8I   
13          13  1sFftb5Y_iY   
14          14  V97utvupGA8   
15          15  O5ukHoPwFAA   
16          16  xdaOoAFIi2M   
17          17  -d0yJQUHQQ0   
18          18  N7FGAHcaDR8   
19          19  pbD4vdj-Vyw   
20          20  Y-QbFBAI_Fo   
21          21  QZeoMHlwBUQ   
22          22  oaDxnmE8juE   
23          23  67Q3YSENx1Y   
24          24  mQBvASRho7Q   
25          25  xUVOmuO3Y7g   
26          26  2gkVOt3Lrvk   
27          27  G2kMU9anaeo   
28          28  fAFld03clSE   
29          29  Z923_dJeBOU   
30          30  wMpT3kkrKHQ   
31          31  c94F-_EzEig   
32          32  w6vCF0DbBsE   
33          33  CzvJ1aaGkSU   
34          34  m93UGg3Unbw   
35          35  WrkmuUf56ow   
36          36  xu_sAM-BnOI   
37          37  VUWDSGdZaB4   
38          38  uaJZp-Va3NI   
39          39  CGN4jH5rir4   
40          40  OKeyoDQ1z6g   
41          41  Q1MPuXyYlMM   
42          42  XRnReOrySGw   
43          43  AglLAB-K1xg   
44          44  H895bJUCues   
45          45  Abi0w98MaEw   
46          46  Jj25RDuBc2g   
47          47  6nsxsHVLbco   
48          48  Gf9RlABpwRU   
49          49  WOvtPpQSpYE   
50          50  7aLqIpxvny0   
51          51  el-20sun0UA   
52          52  1dBGZUoqbjo   
53          53  Ux2GHC9QXnM   
54          54  NMKj0BVJono   
55          55  jhgNSBzzvzQ   
56          56  h7IeZ4qTdFo   
57          57  lrxRJboEGrw   
58          58  WKB2IQz-rcs   
59          59  kCgV4Bl72mY   
60          60  oVgEgVuK2Y8   
61          61  oXzjaCjxcIs   
62          62  y4ZqKsTQN8I   
63          63  1sFftb5Y_iY   
64          64  V97utvupGA8   
65          65  O5ukHoPwFAA   
66          66  xdaOoAFIi2M   
67          67  -d0yJQUHQQ0   
68          68  N7FGAHcaDR8   
69          69  pbD4vdj-Vyw   
70          70  Y-QbFBAI_Fo   
71          71  QZeoMHlwBUQ   
72          72  oaDxnmE8juE   
73          73  67Q3YSENx1Y   
74          74  mQBvASRho7Q   
75          75  xUVOmuO3Y7g   
76          76  2gkVOt3Lrvk   
77          77  G2kMU9anaeo   
78          78  fAFld03clSE   
79          79  Z923_dJeBOU   
80          80  wMpT3kkrKHQ   
81          81  c94F-_EzEig   
82          82  w6vCF0DbBsE   
83          83  CzvJ1aaGkSU   
84          84  m93UGg3Unbw   
85          85  WrkmuUf56ow   
86          86  xu_sAM-BnOI   
87          87  VUWDSGdZaB4   
88          88  uaJZp-Va3NI   
89          89  CGN4jH5rir4   
90          90  OKeyoDQ1z6g   
91          91  Q1MPuXyYlMM   
92          92  XRnReOrySGw   
93          93  AglLAB-K1xg   
94          94  H895bJUCues   
95          95  Abi0w98MaEw   
96          96  Jj25RDuBc2g   
97          97  6nsxsHVLbco   
98          98  Gf9RlABpwRU   
99          99  WOvtPpQSpYE   

                                                title  \
0                                         حفلة عرسي 💍   
1   Vlog- هادي ماكنتش دايراها فالحسبان 😪 ماما تصدم...   
2   أستاذة تدافع عن  سمعة زميلتها الراحلة هاجر وتك...   
3             كان غادي يبكي بالفرحة ملي شاف الموطور🥺🔥   
4               DUKE - WALIDIA (Official Music Video)   
5   ملخص مباراة أستون فيلا وباريس سان جيرمان (3-2)...   
6   ملخص مباراة ريال مدريد وآرسنال (1-2) | دوري أب...   
7   هكذا تفاعل فوزي لقجع مع ضربات الجزاء ضد منتخب ...   
8   اول روتين في دار محبوبة🥳كواليس تحضيرات استقبال...   
9   وأخير 👌حتى انا  درت 🎁لكدو نتاعي لتيتيش  متصورو...   
10  ملخص مباراة إنتر وبايرن ميونيخ (2-2) | دوري أب...   
11  تحديت محمد عدنان في تحدي كروي مجنووون 🤯⚽ | تحد...   
12                  🟦 🟪 اكبر لعبة ألوان مع اليوتيوبرز   
13               ردة فعلي على ميمز عرس نارين بيوتي !!   
14  I'M BACK - رج

### Яка максимальна кількість коментарів серед усіх цих відео?

In [14]:
max_comments_count = np.max(the_most_popular_videos["comment_count"])
int(max_comments_count)

79357

Максимальна кількість коментарів серед усіх відео становить 79357.

### Обчисліть коефіцієнт кореляції між кількістю переглядів і кількістю вподобайок?

In [22]:
view_count = the_most_popular_videos["view_count"]
like_count =  the_most_popular_videos["like_count"]

correlation_matrix = np.corrcoef(view_count, like_count)
correlation_coeff = correlation_matrix[0][1]
float(correlation_coeff)

0.8300126316982337

В pandas:

In [25]:
correlation = view_count.corr(like_count)
float(correlation)

0.8300126316982337

Коефіцієнт кореляції між кількістю переглядів і кількістю вподобайок становить 0.83, що вказує на сильну позитивну кореляцію між цими двома змінними. Це означає, що відео з більшою кількістю переглядів, ймовірно, мають також велику кількість вподобайок.

### З якої категорії найбільше відео серед найпопулярніших?

In [34]:
grouped_by_category = the_most_popular_videos.groupby("category_id")["video_id"].count()
grouped_by_category = grouped_by_category.rename("video_count").reset_index()
grouped_by_category

category_id  video_count
0           10           14
1           17           12
2           20           14
3           22           24
4           23            2
5           24           20
6           25            4
7           26            8
8           27            2

In [35]:
max_value = np.max(grouped_by_category["video_count"])
row = grouped_by_category[grouped_by_category["video_count"] == max_value]
category_id = row["category_id"].iloc[0]
int(category_id)

22

Серед найпопулярніших відео категорія з ідентифікатором 22 містить найбільше відео.

### В якій категорії відео середня кількість коментарів є найбільшою?

In [38]:
grouped_by_category = the_most_popular_videos.groupby("category_id")["comment_count"].mean()
grouped_by_category = grouped_by_category.rename("comment_count_mean_value").reset_index()
grouped_by_category

category_id  comment_count_mean_value
0           10               2492.000000
1           17               4334.833333
2           20              24092.142857
3           22               1433.083333
4           23               1694.000000
5           24               5147.200000
6           25                443.500000
7           26               1713.250000
8           27               4110.000000

In [42]:
max_value = np.max(grouped_by_category["comment_count_mean_value"])
row = grouped_by_category[grouped_by_category["comment_count_mean_value"] == max_value]
int(row["category_id"].iloc[0])

20

Середня кількість коментарів є найбільшою у категорії з ідентифікатором 20, що свідчить про вищу активність користувачів щодо відео в цій категорії.

### О котрій годині (не потрібен точний час з хвилинами) найчастіше публікують популярні відео?

In [46]:
the_most_popular_videos["published_at"] = pd.to_datetime(the_most_popular_videos["published_at"])
the_most_popular_videos["publishing_hour"]= the_most_popular_videos["published_at"].dt.hour
grouped_by_publishing_hour = the_most_popular_videos.groupby("publishing_hour")["publishing_hour"].count().rename("frequency").reset_index()
grouped_by_publishing_hour

publishing_hour  frequency
0                 3          2
1                 9          2
2                10          8
3                12          2
4                13         10
5                14          6
6                15         20
7                16          6
8                17          8
9                18          2
10               19         14
11               20          4
12               21          8
13               22          6
14               23          2

In [49]:
max_value = np.max(grouped_by_publishing_hour["frequency"])
row = grouped_by_publishing_hour[grouped_by_publishing_hour["frequency"] == max_value]
int(row["publishing_hour"].iloc[0])

15

Популярні відео найчастіше публікуються близько 15-ї години, що вказує на тенденцію до активності в цей період для публікації відео.